deepcopy: add alias support

sttts opened the pull request:
~~TODO:~~
- [x] verify that deepcopies in Kubernetes are correct.
- [x] verify that deepcopies in Openshift are correct.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/gengo/73)
<!-- Reviewable:end -->

sttts commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -762,6 +764,8 @@ func (g *genDeepCopy) doSlice(t *types.Type, sw *generator.SnippetWriter) {
 			sw.Do("}\n", nil)
 		} else if t.Elem.Kind == types.Struct {
 			sw.Do("(*in)[i].DeepCopyInto(&(*out)[i])\n", nil)
+		} else if t.Elem.Kind == types.Alias {
```
> thockin:
> a) this should just be an ||
> b) if it is an Alias, you still need to check the underlying Kind, no?

smarterclayton replied:
If you get an alias, the alias needs to be to a struct.

sttts commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -762,6 +764,8 @@ func (g *genDeepCopy) doSlice(t *types.Type, sw *generator.SnippetWriter) {
 			sw.Do("}\n", nil)
 		} else if t.Elem.Kind == types.Struct {
 			sw.Do("(*in)[i].DeepCopyInto(&(*out)[i])\n", nil)
+		} else if t.Elem.Kind == types.Alias {
```
@thockin I guess you are right. On the other hand we also fall back to `.DeepCopy` one case below. My change does not change much here, only optimized a bit. Struct aliases will work with this. For the full solution we need another case. Or fall back to `g.generateFor(t, sw)` with another `{ in, out := ... }` scope.

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -309,6 +309,30 @@ func hasDeepCopyMethod(t *types.Type) bool {
 	return false
 }
 
+// hasDeepCopyMethod returns true if an appropriate postDeepCopy(out *T) method is
```
comment doesn't match function name

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -309,6 +309,30 @@ func hasDeepCopyMethod(t *types.Type) bool {
+// finished.
+//
+//  The correct signature for a type T is:
+//    func (in *T) postDeepCopy(out *T)
```
it feels dirty to be looking at private methods.  Should this be public?

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -770,7 +796,7 @@ func (g *genDeepCopy) doStruct(t *types.Type, sw *generator.SnippetWriter) {
 			if hasMethod {
 				sw.Do("out.$.name$ = in.$.name$.DeepCopy()\n", args)
-			} else if t.IsAssignable() {
+			} else if !hasPostDeepCopyMethod && t.IsAssignable() {
```
why does this care about the method?  Please comment these - it's really twisty

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -471,6 +475,10 @@ func (g *genDeepCopy) needsGeneration(t *types.Type) bool {
 	return true
 }
 
+func tagValueIsTrue(val string) bool {
```
This seems like something that will surprise me later.  Can you at least comment why "" is considered true here?  Better, rename it to be clear

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -494,15 +502,29 @@ func extractNonPointerInterfaces(comments []string) (bool, error) {
+	}
+	result := tagValueIsTrue(values[0])
+	for _, v := range values {
+		if tagValueIsTrue(v) && values[0] != "" {
```
I don't understand why this is comparing to [0] ?

thockin commented on examples/deepcopy-gen/main.go:
```diff
@@ -22,7 +22,10 @@ limitations under the License.
 // generating new ones.
 //
 // If a "func (in *T) postDeepCopy(out *T)" func is predefined on an object by the
```
Why can't this logic just check for and call the T.DeepCopy()?  If the use case is to not generate code and then call custom code, that is exactly what defining DeepCopy does.

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -356,7 +357,7 @@ func copyableType(t *types.Type) bool {
 	}
 	// TODO: Consider generating functions for other kinds too.
-	if t.Kind != types.Struct {
+	if t.Kind != types.Struct && t.Kind != types.Alias {
```
I really need you to explain what this is doing and why?  Now we deepcopy other types beside struct?  I'm OK with that but why Alias?

smarterclayton replied:
You should probably be doing the lookup here to ensure the alias is to a struct.

sttts commented:
> Remind me what the need for this is? I'm liking it less and less as I think about it and the precedent it sets.

Mostly for completeness actually, as you can define funcs on aliases (DeepCopy or postDeepCopy). We don't require it anyway. That's why I moved it out. Am not a fan of the complexity either.

sttts commented:
@thockin "just for completeness" turned into a real use-case, or even a number of them in the OpenShift types. I rebased and fixed a couple of issues.

soltysh commented:
I've double checked it, since I hit this while working on 1.8.1 rebase into OpenShift. This fix is properly addressing aliases in deep copies generation.

deads2k commented:
@sttts rebase please

sttts commented:
> rebase please

done

sttts commented:
@soltysh ptal. In every of the `do*` funcs the given type can be an alias. The same is true for the value of a slice or map. We always pass down the alias type, but do the actual copy code according to the underlying type. There are no changes in Kube with this change.

deads2k commented:
test failures

sttts commented:
/retest

sttts commented:
> test failures

They look infra related.

sttts commented:
@soltysh doublechecked that origin deepcopies look good. They do!

sttts commented:
Tests are not green because of https://github.com/kubernetes/gengo/pull/91

soltysh commented on examples/deepcopy-gen/output_tests/aliases/doc.go:
```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Should probably be 2018 by now ;)

sttts replied:
We don't update unchanged files, do we?

sttts replied:
fixed

soltysh commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -738,34 +770,32 @@ func (g *genDeepCopy) doInterface(t *types.Type, sw *generator.SnippetWriter) {
 	sw.Do("if *in == nil { *out = nil } else {\n", t)
-	if hasDeepCopyMethod(t.Elem) {
-		sw.Do("*out = new($.Elem|raw$)\n", t)
+	if hasDeepCopyMethod(ut.Elem) {
```
I think it's worth the comment you left me [here](https://github.com/kubernetes/gengo/pull/73#issuecomment-354788061) why we check `ut.Elem` but generate for `uet`. It's something that needs to be cleared to everyone reading the code.

sttts replied:
done

soltysh commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -578,93 +618,96 @@ func (g *genDeepCopy) doBuiltin(t *types.Type, sw *generator.SnippetWriter) {
 			sw.Do("(*out)[key] = val\n", nil)
 			sw.Do("}\n", nil)
-		case t.Elem.Kind == types.Interface:
+		case ut.Elem.Kind == types.Interface: // only call this for non-aliased interfaces
```
If `ut.Elem` is an interface then `underlyingType` won't change it. Why not using `uet.Kind` here in that case?

sttts replied:
Actually this does not matter. The golang parser renders interface aliases as interfaces, without access to the underlying type. I fixed this, added comments and examples.

soltysh commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -578,93 +618,96 @@ func (g *genDeepCopy) doBuiltin(t *types.Type, sw *generator.SnippetWriter) {
 			//} else if t.Elem.IsAssignable() {
 			//	sw.Do("(*out)[i] = (*in)[i]\n", nil)
-		} else if t.Elem.Kind == types.Interface {
+		} else if ut.Elem.Kind == types.Interface {
```
Same here, wrt `uet.Kind` instead of `ut.Elem.Kind`?

sttts replied:
dito

soltysh commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -683,46 +728,33 @@ func (g *genDeepCopy) doStruct(t *types.Type, sw *generator.SnippetWriter) {
 				sw.Do("in.$.name$.DeepCopyInto(&out.$.name$)\n", args)
 			}
-		case types.Interface:
+		case t.Kind == types.Interface: // only call this for non-aliased interfaces
```
`uft.Kind` ?

sttts replied:
dito

sttts commented:
@soltysh addressed your comments. ptal.

soltysh commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -561,110 +603,127 @@ func (g *genDeepCopy) generateFor(t *types.Type, sw *generator.SnippetWriter) {
 				sw.Do("val.DeepCopyInto(newVal)\n", nil)
 				sw.Do("(*out)[key] = *newVal\n", nil)
-			} else if t.Elem.Kind == types.Slice && t.Elem.Elem.Kind == types.Builtin {
+			} else if uet.Kind == types.Slice || uet.Kind == types.Builtin {
```
Should be `&&`.

sttts replied:
fixed

soltysh commented:
/assign @thockin for merge

sttts commented:
Test fail because of https://github.com/kubernetes/gengo/pull/91.

thockin commented on examples/deepcopy-gen/output_tests/wholepkg/zz_generated.go:
```diff
@@ -761,9 +763,7 @@ func (in *Struct_Slices_Alias) DeepCopyInto(out *Struct_Slices_Alias) {
 			(*out)[i] = (*in)[i].DeepCopy()
 		}
 	}
-	if in.ManualSliceField != nil {
```
These deltas are suspicious...  Is it right?  Why?

thockin commented:
Can you update the first comment in the PR (the "main" comment) to describe what this is doing and why?

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -292,21 +292,40 @@ func (g *genDeepCopy) copyableAndInBounds(t *types.Type) bool {
+	return true
+}
+
+// hasDeepCopyIntoMethod returns true if an appropriate DeepCopyInit() method is
```
Init() or Into() ?

sttts replied:
Into()

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -292,21 +292,40 @@ func (g *genDeepCopy) copyableAndInBounds(t *types.Type) bool {
+	if len(f.Signature.Parameters) != 1 {
+		return false
+	}
+	if len(f.Signature.Results) != 0 {
```
shouldn't this also be checking the type of the argument (kind == pointer, etc)?

sttts replied:
Actually, I tend to remove all these checks. IMO it's better to create code that is not type-correct than skipping a DeepCopy functions that the user expects to be called, but it isn't because there is slight type-error. If we don't want to rely on the type checker, we could also error out here in the generator. But skipping slightly off funcs leads to hard to understand wrong behaviour of the generated code.

sttts replied:
I added another commit with fatal error. I would very much prefer that. If this turns out to be a problem on any external type with DeepCopy/Into, we can still add an opt-out tag. But we haven't seen such a case yet, so let's keep it simple for now.

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -327,17 +346,38 @@ func copyableType(t *types.Type) bool {
+		if hasDeepCopyMethod(t) || hasDeepCopyIntoMethod(t) {
+			return true
+		} else {
+			return t.Underlying.Kind != types.Builtin || copyableType(t.Underlying)
```
won't this catch Map and Slice when it didn't used to?

sttts replied:
Yes, it does. It's intentional, compare https://goplay.space/#KWig91IjuQv

Will add a comment.

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -327,17 +346,38 @@ func copyableType(t *types.Type) bool {
 }
 
+func underlyingType(t *types.Type) *types.Type {
+	for {
```
```
for t.Kind == types.Alias {
    t = t.Underlying
}
```
 
?

sttts replied:
:+1:

thockin commented on examples/deepcopy-gen/output_tests/aliases/doc.go:
```diff
@@ -0,0 +1,89 @@
+	DeepCopyAliasAliasInterface() AliasAliasInterface
+}
+
+type Builtin int
```
It would actually be way nicer to review / maintain a couple aliases added to each of the existing output_tests, rather than this.

e.g. aliases for int/string in output_tests/builtins, etc

sttts replied:
I think both ways have advantages, depending on your view point. To be complete with alias support I pretty much prefer this structure. We can maintain both if you feel strongly about this.

thockin commented:
Does this produce deltas against kubernetes?

sttts commented:
> Does this produce deltas against kubernetes?

Noticed an issue with aliased pointers, maps and slices and pre-defined DeepCopy funcs. On it....

sttts commented:
@thockin here is the diff of kube. Looked through most of the changes and they look. But more eyes on the changes would be good. https://gist.github.com/sttts/7474ed37382b7abc79eeaeb99a9929b6

jennybuckley commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -286,27 +286,76 @@ func (g *genDeepCopy) copyableAndInBounds(t *types.Type) bool {
+// for a type T is:
+//    func (t T) DeepCopyInto(t *T)
+// or:
+//    func (t *T) DeepCopyInit(t *T)
```
Won't this return nil for "func (t *T) DeepCopyInit(t *T)" ?

sttts replied:
typo. fixed.

soltysh commented:
From notable changed the following types got `DeepCopyInto` methods:

* - `ExtraValue`
* - `ResourceList`
* - `SortableRuleSlice`
* - `ConfigurationMap`
* - `FailedNodesMap`
* - `HostPriorityList`
* - `JSONSchemaDefinitions`
* - `JSONSchemaDependencies`
* - `Verbs`
* - `ByGroupPriorityMinimum`

The few remaining changes look sane.

sttts commented:
> ExtraValue

This was actually wrong before, turning `nil` into a zero element slice.

> ResourceList

It's an alias and the deepcopy funcs are correct.

> SortableRuleSlice

It's an alias and the deepcopy funcs are correct.

> ConfigurationMap

It's an `map[string]string` alias and the deepcopy funcs are correct.

> FailedNodesMap

It's an `map[string]string` alias and the deepcopy funcs are correct.

> HostPriorityList

It's a struct slice alias with copyable structs, and the deepcopy funcs are correct.

> JSONSchemaDefinitions

It's a non-copyable JSONSchemaProps string map alias, and the deepcopy funcs are correct.

> JSONSchemaDependencies

It's a non-copyable JSONSchemaPropsOrStringArray string map alias, and the deepcopy funcs are correct.

> Verbs

It's a string slice alias, and the deepcopy funcs are correct.

> ByGroupPriorityMinimum

It's a non-copyable `*APIService` slice alias and the deepcopy funcs are correct.

sttts commented:
In short: we get a couple of alias `DeepCopy` funcs. But's that's intended.

soltysh commented:
SGTM, ship it!

sttts commented:
Rebased.

thockin commented:
PR comment says "TODO".  Please update.

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -286,27 +286,76 @@ func (g *genDeepCopy) copyableAndInBounds(t *types.Type) bool {
 	}
-	return false
+	if len(f.Signature.Parameters) != 0 {
+		glog.Fatalf("Type %v: invalid DeepCopy signature, expected no parameters", t)
```
Why are we aborting?  Other tools just ignore signatures that don't match.  e.g.

https://play.golang.org/p/tl05gzSswY-

sttts replied:
Having a wrong signature is most probably a programmer error. And assuming the custom deepcopy being called, but isn't, sound dangerous and hard to debug. I prefer to fail early here.

soltysh replied:
I second that opinion. Chasing bugs like that, which can be easily caught earlier is definitely better option.

thockin commented on examples/deepcopy-gen/generators/deepcopy.go:
```diff
@@ -286,27 +286,76 @@ func (g *genDeepCopy) copyableAndInBounds(t *types.Type) bool {
+//    func (t T) DeepCopyInto(t *T)
+// or:
+//    func (t *T) DeepCopyInto(t *T)
+func deepCopyIntoMethod(t *types.Type) *types.Signature {
```
should we perhaps demand that a user defines both of these or neither of these?  Or are we ok to mix-and-match?

sttts replied:
We are ok. One alone works for us. The DeepCopyInto is usually an optimization with less allocations.

thockin commented:
Travis fail:  examples/deepcopy-gen/generators/deepcopy_test.go:263: undefined: hasDeepCopyMethod

thockin commented:
/approve

sttts commented:
> Travis fail: examples/deepcopy-gen/generators/deepcopy_test.go:263: undefined: hasDeepCopyMethod

Fixed tests and completed them with new cases and deepCopyInto tests.

thockin commented:
/approve

sttts commented:
Thanks for merging. Updating the kube PR to revendor it.
